controllers/etcdrestore_controller.go

```diff
@@ -0,0 +1,524 @@
+		pod.Labels[restorePodLabel] == "true"
+}
+
+func markCluster(restore etcdv1alpha1.EtcdRestore, cluster *etcdv1alpha1.EtcdCluster) {
+	if cluster.Labels == nil {
+		cluster.Labels = make(map[string]string)
+	}
+	cluster.Labels[restoredFromLabel] = restore.Name
+}
+
+func (r *EtcdRestoreReconciler) podForRestore(restore etcdv1alpha1.EtcdRestore, pvc *corev1.PersistentVolumeClaim) *corev1.Pod {
+	return &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      restore.Name,
+			Namespace: restore.Namespace,
+			Labels: map[string]string{
+				restoredFromLabel: restore.Name,
+				restorePodLabel:   "true",
+			},
+		},
+		Spec: corev1.PodSpec{
+			Volumes: []corev1.Volume{
+				{
+					Name: "etcd-data-directory",
+					VolumeSource: corev1.VolumeSource{
+						PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc.Name,
+							// We definitely need to be able to write to it.
+							ReadOnly: false,
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:         "etcd-restore",
+					Image:        r.RestorePodImage,
+					Command:      nil,
+					Args:         nil,
+					VolumeMounts: nil,
+				},
+			},
+			// If the Pod fails we should *not* attempt to recover automatically. Leave it failed.
+			RestartPolicy: corev1.RestartPolicyNever,
+		},
+	}
+}
+
+func (r *EtcdRestoreReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("etcdrestore", req.NamespacedName)
+
+	log.Info("Begin restore reconcile")
+
+	var restore etcdv1alpha1.EtcdRestore
+	if err := r.Get(ctx, req.NamespacedName, &restore); err != nil {
+		// Can't find the resource. We could have just been deleted? If so, no need to do anything as the owner
+		// references will clean everything up with a cascading delete.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	} else {
+		log.Info("Found restore object in Kubernetes, continuing")
+		// Found the resource, continue with reconciliation.
+	}
+
+	if restore.Status.Phase == etcdv1alpha1.EtcdRestorePhaseCompleted ||
+		restore.Status.Phase == etcdv1alpha1.EtcdRestorePhaseFailed {
+		// Do nothing here. We're already finished.
+		log.Info("Phase is set to an end state. Taking no further action", "phase", restore.Status.Phase)
+		return ctrl.Result{}, nil
+	} else if restore.Status.Phase == etcdv1alpha1.EtcdRestorePhasePending {
+		// In any other state continue with the reconciliation. In particular we don't *read* the other possible states
+		// as once we know that we need to reconcile at all we will use the observed sate of the cluster and not our own
+		// status field.
+		log.Info("Phase is not set to an end state. Continuing reconciliation.", "phase", restore.Status.Phase)
+	} else {
+		log.Info("Restore has no phase, setting to pending")
+		restore.Status.Phase = etcdv1alpha1.EtcdRestorePhasePending
+		err := r.Client.Status().Update(ctx, &restore)
```

Comment:
In ClusterAPI project, they set update the Status as part of a deferred function, which I quite like:
 * https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/machine_controller.go#L127-L135

This way we always attempt to update the status and we only have to do it in one place.
Consider it.

Reply:
Right now the status setting is in-with the main reconciliation because it's tightly tied to it. If there's no phase it should be pending, if it's pending it should restore normally, if it's failed or completed it should exit.

Then the phase is set based on what happens. Setting to failed if the pods die, and setting to completed if it reaches the end and has a cluster.

I'm not sure how I could pull it out into a deferred function. It would need to know information from the reconcile before it to work anyway, so I don't think it would do anything except require indirection in shared variables or channels to pass around knowledge of what to set the status to.

Am I misunderstanding?